File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.java
import java.util.*;
class Solution {
    public int[] solution(int n) {
        int[][] nums = new int[n][n];
        ArrayList<Integer> list = new ArrayList<>();
        int num = 1;
        int x = -1, y = 0;
        for(int i = 0; i < n; i++) {
           for(int j = i; j < n; j++) {
               if(i % 3 == 0) {
                   x++;
               }
               else if (i % 3 == 1) {
                   y++;
               }
               else if(i % 3 == 2) {
                   x--;
                   y--;
               }
               nums[x][y] = num++;
           }
        }
        for(int i = 0; i < nums.length; i++) {
            for(int j = 0; j <= i; j++) {
                list.add(nums[i][j]);
            }
        }
        int[] answer = new int[list.size()];
        for(int i = 0; i < list.size(); i++) {
            answer[i] = list.get(i);
        }
        
        return answer;
    }
}

File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.js
function solution(n) {
    var answer = [];
    // 몫만큼 반복,
    const arr = Array(n).fill().map((_,i) => Array(i+1).fill(0));
    const sum = Array(n).fill().map((_,i) => i+1).reduce((acc,cur)=>acc+cur,0);
    const dx = [0,1,-1];
    const dy= [1,0,-1];
    
    let num = 1;
    let index = 0;
    let x = 0, y = 0;
    while (num <= sum) {
        arr[y][x] = num++;
        let ny = y + dy[index];
        let nx = x + dx[index];
       if (ny < 0 || ny >= arr.length || nx < 0 || nx >= arr[ny].length || arr[ny][nx] !== 0) {
           index = (index + 1) % 3;
           ny = y + dy[index];
           nx = x + dx[index];
        } 
        y = ny;
        x = nx;
        
    }
    return arr.flat();
}